src/nodes/hypertable_modify.c

```diff
@@ -93,6 +95,287 @@ get_chunk_dispatch_states(PlanState *substate)
 	return NIL;
 }
 
+#if PG14_GE
+/*
+ * Helper method which returns true if SEGMENT BY column
+ * is NULLABLE. This method also returns the attribute
+ * number of the SEGMENT BY column in compressed chunk.
+ */
+static bool
+is_segmentby_col_null(Chunk *curr_chunk, char *column_name, int *attrno)
+{
+	Chunk *comp_chunk = NULL;
+	Relation chunk_rel;
+	Relation comp_chunk_rel;
+	TupleDesc tupleDesc;
+	int attno = 0;
+	bool is_null = false;
+
+	chunk_rel = relation_open(curr_chunk->table_id, AccessShareLock);
+	tupleDesc = RelationGetDescr(chunk_rel);
+	/*
+	 * check if SEGMENT BY column is NULLABLE or not based on uncompressed
+	 * chunk tuple descriptor. This is because compressed chunk column
+	 * attributes don't have NULLABLE properties.
+	 */
+	for (attno = 1; attno <= tupleDesc->natts; attno++)
+	{
+		Form_pg_attribute attribute = TupleDescAttr(tupleDesc, AttrNumberGetAttrOffset(attno));
+		if (attribute->attisdropped)
+			continue;
+		char *attributeName = NameStr(attribute->attname);
+		if (strcmp(attributeName, column_name) == 0)
+		{
+			*attrno = attno;
+			is_null = attribute->attnotnull;
+			break;
+		}
+	}
+	table_close(chunk_rel, AccessShareLock);
+	/*
+	 * we get attrno for SEGMENT BY column from uncompressed chunk. This
+	 * can be different from that of compressed chunk due to different
+	 * physical layouts of compressed and uncompressed chunks. Thus we
+	 * need to retreive correct attrno from compressed chunk tuple descriptor.
+	 */
+	comp_chunk = ts_chunk_get_by_id(curr_chunk->fd.compressed_chunk_id, true);
+	comp_chunk_rel = relation_open(comp_chunk->table_id, AccessShareLock);
+	tupleDesc = RelationGetDescr(comp_chunk_rel);
+	for (attno = 1; attno <= tupleDesc->natts; attno++)
+	{
+		Form_pg_attribute attribute = TupleDescAttr(tupleDesc, AttrNumberGetAttrOffset(attno));
+		if (attribute->attisdropped)
+			continue;
+		char *attributeName = NameStr(attribute->attname);
+		if (strcmp(attributeName, column_name) == 0)
+		{
+			*attrno = attno;
+			break;
+		}
+	}
+	table_close(comp_chunk_rel, AccessShareLock);
```

Comment:
This function should be rewritten to use the postgres catalog cache instead, this would simplify the code a lot. Have a look at get_attnum for an example.

Reply:
Thanks, fixed it.